### ISSUE TITLE
RegisterRoutes is getting called twice

### ISSUE THREAD

**Issue by mobynet1** [unoplatform/uno#18121]
### Current behavior

I placed some code in the `RegisterRoutes` method in `App.xaml.cs` to check for the existence of a control file to determine the correct first route to take on app startup. I placed a breakpoint in the method just to make sure my code did what I wanted it to do, and the breakpoint was hit twice. I could not tell from the call stack why this is happening. It is being called by the `.UseNavigation` extension method, but I could not find any other point where it is being called. I ran a new uno solution that I just created and it did not hit the breakpoint twice.

I suspect I know why. Here is some code from the `OnLaunched` method:

```text
	MainWindow.SetWindowIcon ();

	Host = builder.Build ();

	_ = await builder.NavigateAsync<Shell> ();
```

I am doing this because I found that Host was not being set without the call in the middle of the above. I need the Host to be accessible from other points in my application so I can get objects from the DI container. If I don't make this first call to the `builder.Build` method, Host is not getting set. If I don't call the `builder.NavigateAsync` method the `INavigator navigator` is not getting injected as expected.

### Expected behavior

The RegisterRoutes method should only be called once.

### Workaround

Live with it for now, but I don't think this is wise.

### Works on UWP/WinUI

None

### Environment

Uno.UI / Uno.UI.WebAssembly / Uno.UI.Skia

### Affected platforms

Windows (WinAppSDK)

### IDE

Visual Studio 2022

### IDE version

Visual Studio 2022 Version 17.11.2

### Anything else we need to know?

nope...

**Comment by mobynet1** [comment]
I figured it out... Doh!

**Comment by agneszitte** [comment]
> I figured it out... Doh!

Hi @mobynet1, I am curious, documentation was missing some details in your opinion? (cc @kazo0, @eriklimakc)

**Comment by mobynet1** [comment]
@agneszitte Well, this is sort of embarrassing, but...

I thought I had tried to get the host simply with the `Host = await builder.NavigateAsync<Shell> ();` call, yet when I was trying to use dependency injection, which relies on the Host object to get at the service collection, I was not seeing the Host object in my view models. I don't know how I had things configured at that time, but even with the `Host` object being made `public`, it was null when I looked at it. When I changed the code to read as I indicated in my initial post, that is when I started to see the `RegisterRoutes` method being called twice. I was probably changing too many things at once trying to get DI to work; the `INavigator` wasn't getting injected; `Host` was `null`, the view models were not visible... The usual things happening to a novice Uno developer who refuses to read the documentation! When I slowed down and took things more piecemeal, I discovered that the injection of the `INavigator` was dependent upon making the call `Host = await builder.NavigateAsync<Shell> ();`. With that, I had the navigator injection issue taken care of. Then, I simply removed the extra `Host = builder.Build ();` call, and the double call to the `RegisterRoutes` method behaved as expected. And, I must have had something set, and I don't remember what, that was preventing `Host` from being visible, but after I cleaned things up and went back to the default `App.xaml.cs` configuration, I started to see `Host` in my view models. I also set a `DataContextChanged` event handler in one of my pages to watch how it was being set. It looks like it progressed like this: null -> ShellViewModel -> xxxViewModel`. In a reply to a discussion question someone indicated the Uno team was going to change how that worked so the data context would not be set, then reset.

(Dog panting noise here...) The world started to rotate in the correct direction and I learned some things along the way!! :-)
